### PR TITLE
New segment filters

### PR DESCRIPTION
**Pull request by kuzmany**
| Q  | A
| --- | ---
| Bug fix? | n
| New feature? | y
| Related user documentation PR URL | n
| Related developer documentation PR URL | n
| Issues addressed (#s or URLs) | #3067
| BC breaks? | n
| Deprecations? | n

#### Description:

This PR added new filters. 
- Visited URL Date - best use case is with Visited URL filter  If I want filter contacts just from some period
- Visited URL Count - filter contacts based on Page Hit Count
- Referer - same like Visited URL  just based on Referer
- URL title - same like Visited URL  just based on URL title
- Source - same like Visited URL  just based on source (email, forms, campaign, dynamic content)
- Email Read Date - best use case is with Email Read filter  If I want filter contacts just from some period
- Email Read Count - filter contacts based on email open_counts
- Notifications - filter for segments by notifications state based on push_leads table
- Email sent - contact list from sent email(s)
- Email URL, Redirect URL, Landing page URL check
- Sessions - visits with 30 minutes between  return - same metric like Google Analytics

#### Steps to test this PR:
1. Apply PR, Go to segments and create new segments
2.  Use Visited URL and Visited URL Date filter combination
3. Call mautic:segments:rebuild and check rebuilded database
4. Create another segment with Referer filter
5. Call mautic:segments:rebuild and check rebuilded database
6. Create another segment with Email Read  and Email Read Date filter
7. Call mautic:segments:rebuild and check rebuilded database
6. Create another segment with Notification filter - choose empty (hasn't notifications) or notEmpty (has notifcation)
7. Call mautic:segments:rebuild and check rebuilded database
8. Create another segment with Visited URL Count 
9. Call mautic:segments:rebuild and check rebuilded database
10. Create another segment with  Email Read Count
11. Call mautic:segments:rebuild and check rebuilded database

### CODE REVIEW

**Comment by MaxWebmecanik** [comment]
Hi @kuzmany, I've been pointed to this PR following my issue #3067 
Does this PR add the ability to filter contacts based on whether or not they were sent an email?

**Comment by kuzmany** [comment]
Not now, but I will add it. Good request.

**Comment by kuzmany** [comment]
@MaxWebmecanik Email sent filter added.

**Comment by MaxWebmecanik** [comment]
@kuzmany Great! Thanks a lot! :smiley:

**Comment by elfeffe** [comment]
I did a checkout on the PR

git checkout pr/2912

And I get this error

PHP Fatal error:  Class 'Mautic\LeadBundle\EventListener\ButtonSubscriber' not found in /htdocs/mautic/vendor/symfony/event-dispatcher/ContainerAwareEventDispatcher.php on line 152

File: app/bundles/LeadBundle/Entity/LeadListRepository.php [hunk 1]
```diff
@@ -855,33 +855,94 @@ public function getListFilterExpr($filters, &$parameters, QueryBuilder $q, $not
...
                 case 'hit_url':
-                    $operand = (($func == 'eq') || ($func == 'like')) ? 'EXISTS' : 'NOT EXISTS';
-
+                case 'hit_url_date':
+                case 'referer':
+                case 'source':
+                case 'url_title':
+                case 'lead_email_received_date':
+                case 'notification':
+                    $operand = (($func == 'notEmpty') || ($func == 'between') || ($func == 'eq') || ($func == 'like') || ($func == 'gt') || ($func == 'lt')) ? 'EXISTS' : 'NOT EXISTS';
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Entity/LeadListRepository.php]: I guess, here `in_array` should be used. And also there might be other cases like `lte` an `gte`. So my suggestion is to turn it into
```
$operand = in_array($func, ['notEmpty', 'between', 'eq', 'like', 'gt', 'gte', 'lt', 'lte']) ? 'EXISTS' : 'NOT EXISTS';
```

**Comment by PatchRanger** [comment]
@kuzmany I've applied the PR at my Mautic installation - and it looks great, thank you for it! I have few suggestions:
- I think `lead_email_received_date` has bad wording, `received` is confusing. I know, it is not introduced by your PR, but I suggest to think about renaming to `lead_email_read_date`. Will it backward compatible to rename the filter?
- I think, it is better to split all date filters into 2 separate filters: `*_first` and `*_last` - because it is not obvious what kind of date is compared. Your current date filters match `*_first` but I failed to realize it until I looked directly into the code: I set `lead_email_received_date` filter to `-1 month` supposing it means `those, who haven't received(opened) any email within 1 month` - but in fact it means `those, who ever opened any email earlier than 1 month ago`. So I suggest to split: `lead_email_received_date` into `lead_email_received_date_first` and `lead_email_received_date_last`, `hit_url_date` into `hit_url_date_first` and `hit_url_date_last`.
- I suggest to also add `lead_email_sent_date` - also in 2 variants: `*_first` and `*_last`.
- Also `lead_email_sent_count` should be added.

File: app/bundles/LeadBundle/Form/Type/FilterType.php [hunk 2]
```diff
@@ -126,6 +126,17 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                     $customOptions['multiple'] = true;
                     $type                      = 'choice';
                     break;
+                case 'lead_email_sent':
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Form/Type/FilterType.php]: As it perfectly fits the previous section - I suggest to just merge them, moving `case 'lead_email_sent':` directly after `case 'lead_email_received':`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 3]
```diff
@@ -342,6 +342,55 @@ public function getChoiceFields()
...
+                        '!in',
+                    ],
+                ],
+                'object' => 'lead',
+            ],
+            'lead_email_received_date' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.lead_email_received_date'),
+                'properties' => ['type' => 'datetime'],
+                'operators' => [
+                    'include' => [
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: Should also include `gte`, `lte`, `empty`, `notEmpty`, `between`, `notBetween`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 4]
```diff
@@ -342,6 +342,55 @@ public function getChoiceFields()
...
+                        'lt',
+                    ],
+                ],
+                'object'     => 'lead',
+            ],
+            'lead_email_received_count' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.lead_email_received_count'),
+                'properties' => ['type' => 'number'],
+                'operators' => [
+                    'include' => [
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: Should also include `=`, `!=`, `gte`, `lte`, `between`, `notBetween`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 5]
```diff
@@ -342,6 +342,55 @@ public function getChoiceFields()
...
+                        'lt',
+                    ],
+                ],
+                'object'     => 'lead',
+            ],
+            'sessions' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.session'),
+                'properties' => ['type' => 'number'],
+                'operators' => [
+                    'include' => [
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: Should also include `!=`, `gte`, `lte`, `between`, `notBetween`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 6]
```diff
@@ -411,6 +460,118 @@ public function getChoiceFields()
                 ],
                 'object' => 'lead',
             ],
+            'referer' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.referer'),
+                'properties' => [
+                    'type' => 'text',
+                ],
+                'operators' => [
+                    'include' => [
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: I think it is worth adding `empty` and `notEmpty`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 7]
```diff
@@ -411,6 +460,118 @@ public function getChoiceFields()
...
+                ],
+                'object' => 'lead',
+            ],
+            'url_title' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.url_title'),
+                'properties' => [
+                    'type' => 'text',
+                ],
+                'operators' => [
+                    'include' => [
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: Should also include `!=`, `!like`, `empty`, `!empty`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 8]
```diff
@@ -411,6 +460,118 @@ public function getChoiceFields()
...
+                ],
+                'object' => 'lead',
+            ],
+            'source' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.source'),
+                'properties' => [
+                    'type' => 'text',
+                ],
+                'operators' => [
+                    'include' => [
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: Should also include `!=`, `!like`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 9]
```diff
@@ -411,6 +460,118 @@ public function getChoiceFields()
...
+                    ],
+                ],
+                'operators' => 'bool',
+                'object'    => 'lead',
+            ],
+            'hit_url_date' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.visited_url_date'),
+                'properties' => ['type' => 'datetime'],
+                'operators' => [
+                    'include' => [
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: Should also include `empty`, `!empty`, `gte`, `lte`, `between`, `!between`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 10]
```diff
@@ -411,6 +460,118 @@ public function getChoiceFields()
...
+                        'lt',
+                    ],
+                ],
+                'object'     => 'lead',
+            ],
+            'hit_url_count' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.visited_url_count'),
+                'properties' => ['type' => 'number'],
+                'operators' => [
+                    'include' => [
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: Should also include `=`, `!=`, `gte`, `lte`, `between`, `!between`.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 11]
```diff
@@ -411,6 +460,118 @@ public function getChoiceFields()
...
+                'object' => 'lead',
+            ],
+            'notification' => [
+                'label'      => $this->translator->trans('mautic.lead.list.filter.notification'),
+                'properties' => [
+                    'type' => 'text',
+                ],
+                'operators' => [
+                    'include' => [
+                        'empty',
```
Comment (PatchRanger) [inline app/bundles/LeadBundle/Model/ListModel.php]: Shouldn't it be boolean instead? Could there be any other values than boolean? If no - it definitely should be bollean))

**Comment by kuzmany** [comment]
@PatchRanger 
Filters supports relative values for dates. Not documented yet but it works like this:
- you can use in your language set in Mautic (obviously english): today, yesterday, tomorrow, last or this or next week/month/year (last year, next month....)

**Filter example for last month on 14. february 2017** 

equal expression is date between 1. 1. 2017 and 31. 1. 2017
gt expression  is  greater then 31. 1. 2017
lt expression  is  less then 1. 1. 2017

**Comment by PatchRanger** [comment]
@kuzmany Thanks for clarification. The problem is not with "-1 month" expression, it works great as well as other relative expressions you mentioned: today, yesterday, tomorrow, last month. The problem is that it does what I don't suppose it to do) The meaning of "filter by Date Received" (or other date filters) is confusing as it could imply 2 different meanings: "filter by **first** Date Received" and "filter by **last** Date Received". That's why I suggest to split them into 2 separate filters each - to avoid confusing misusage.

**Comment by kuzmany** [comment]
@PatchRanger Ok, I got it. I didn't know about relative expression like -1 month and +1 month.  
Thats make sense, now.

**Comment by yourdigitalclub** [comment]
I'd suggest adding a filter that can be used to check if contacts are on any segments and also adding a filter to limit the number of contacts in a segment

**Pull request closed without merging**